File: Java/Linked_List/Reverse_Linked_List_II.java
// https://leetcode.com/problems/reverse-linked-list-ii/
/*
        // Definition for Singly-Linked List
        public class ListNode
	{

                int val;
                ListNode next;

                public ListNode()
                {

                }

                public ListNode( int val )
                {
                        this.val = val;
                }

                public ListNode( int val, ListNode next )
                {
                        this.val = val;
                        this.next = next;
                }

        }
*/
class Solution
{
	public ListNode reverseBetween( ListNode head, int left, int right )
        {
		// WHEN HEAD IS NULL/EMPTY THEN WE RETURN NULL
                if( head == null )
		{
			return null;
		}

		// WHEN LEFT & RIGHT POINTERS ARE EQUAL, THEN NO SWAPPING/REVERSING OF ELEMENTS IS NEEDED
                if( left == right )
                {
                        return head;
                }

                ListNode beforeLinkingNode = head; // HERE beforeLinkingNode IS THE HEAD
		ListNode afterLinkingNode = head.next; // HERE afterLinkingNode IS THE TAIL
		ListNode previous = head;
                ListNode current = head.next;
		short counter = 1;

		while( counter + 1 < left )
		{
                        beforeLinkingNode = beforeLinkingNode.next;
			counter++;
		}

                previous = beforeLinkingNode;
                current = previous.next;
                afterLinkingNode = current;
		right -= counter;

		// REVERSES NODES BETWEEN LEFT & RIGHT POINTER
                while( right > 0 )
		{
                        ListNode next = current.next;
                        current.next = previous;
			previous = current;
			current = next;
			right--;
		}
		/*
			HERE TWO IMPORTANT CASES ARE COVERED
			CASE 1:
				head: [1, 2]
				left: 1
				right: 2
				WHEN left = 1, THEN WE FOLLOW THE NORMAL REVERSING LOGIC, BECASUE beforeLinkingNode WILL ALWAYS POINT TO ITS ORIGINAL POSITION
			CASE 2:
				head: [1, 2, 3]
				left: 1
				right: 3
				WHEN left = 1, THEN WE FOLLOW THE NORMAL REVERSING LOGIC, BECASUE beforeLinkingNode WILL ALWAYS POINT TO ITS ORIGINAL POSITION
		*/
                if( left == 1 )
                {
                        afterLinkingNode = head;
                        afterLinkingNode.next = current;
                        head = previous;
                }
                else
                {
                        beforeLinkingNode.next = previous;
                        afterLinkingNode.next = current;
                }

		return head;

        }
}

File: Linked_List/Reverse_Linked_List/Reverse_Linked_List_II.js
// https://leetcode.com/problems/reverse-linked-list-ii/

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */

/**
 * @param {ListNode} head
 * @param {number} left
 * @param {number} right
 * @return {ListNode}
 */

const reverseBetween = (head, left, right) => {
        if (head == null || left >= right) {
                return head;
        }

        let current = head;
        let leftBefore = head;
        let index = 1;
        while (index < left) {
                leftBefore = current;
                current = current.next;
                index++;
        }

        let tail = current;
        let previous = null;
        // Reverse the Linked List between left & right pointer
        while (index >= left && index <= right) {
                const next = current.next;
                current.next = previous;
                previous = current;
                current = next;
                index++;
        }

        leftBefore.next = previous;
        tail.next = current;

        return left > 1 ? head : previous;
};